observatory_platform/telescopes/crossref_events.py

```diff
@@ -0,0 +1,710 @@
+            batches = [(start_date, end_date)]
+        else:
+            raise AirflowException(f'Download mode has to be either "sequential" or "parallel", '
+                                   f'not "{CrossrefEventsTelescope.DOWNLOAD_MODE}"')
+
+        self.urls = []
+        edited_deleted = False
+        for batch in batches:
+            start_date = batch[0]
+            end_date = batch[1]
+            event_type_urls = [CrossrefEventsTelescope.EVENTS_URL.format(mailto=mail_address, start_date=start_date,
+                                                                         end_date=end_date)]
+            # only add edited/deleted urls to first batch, since no end date can be specified
+            if edited_deleted or first_release:
+                pass
+            else:
+                event_type_urls.append(CrossrefEventsTelescope.EDITED_URL.format(mailto=mail_address,
+                                                                                 start_date=start_date))
+                event_type_urls.append(CrossrefEventsTelescope.DELETED_URL.format(mailto=mail_address,
+                                                                                  start_date=start_date))
+                edited_deleted = True
+            self.urls.append(event_type_urls)
+
+    @property
+    def batch_dates(self) -> list:
+        """
+        Create batches of time periods, based on the start and end date of the release and the max number of
+        processes available.
+
+        :return: List of batches, where each batch is a tuple of (start_date, end_date). Both dates are strings in
+        format YYYY-MM-DD
+        """
+        start_date = self.start_date.date()
+        end_date = self.end_date.date()
+        max_processes = CrossrefEventsTelescope.MAX_PROCESSES
+
+        # number of days between start and end, add 1, because end date is included
+        total_no_days = (end_date - start_date).days + 1
+
+        # number of days in each batch, rounded to nearest integer, minimal is 1
+        batch_no_days = max(1, round(total_no_days / max_processes))
+
+        batches = []
+        period = pendulum.period(start_date, end_date)
+        for i, dt in enumerate(period.range('days', batch_no_days)):
+            batch_start_date = dt.strftime("%Y-%m-%d")
+            # if final batch or end date is reached before end of period, use end date of release instead
+            if i == (max_processes-1) or dt == end_date:
+                batch_end_date = end_date.strftime("%Y-%m-%d")
+                batches.append((batch_start_date, batch_end_date))
+                break
+            # use end date of period
+            else:
+                # end date is included, so subtract 1 day from batch_no_days
+                batch_end_date = (dt+timedelta(days=batch_no_days-1)).strftime("%Y-%m-%d")
+                batches.append((batch_start_date, batch_end_date))
+
+        return batches
```

Comment:
It seems that the date range of each batch changes depending on what start_date and end_date is supplied. For instance, if a date range less than a month is supplied, it will create batches for each day, or if the date range is great enough it will split the batches into months. Would it make it simpler to just keep the batch period the same, except for the last period?

Tuan has created a function called build_schedule that is merged into develop, which you might be able to use here (or extend if you want slightly different functionality): https://github.com/The-Academic-Observatory/observatory-platform/blob/develop/observatory_platform/utils/telescope_utils.py#L36-L53

This type of functionality is good to put into its own function because it makes it easy to test on its own or reuse it.

Reply:
I preferred the use of flexible batch periods, so you optimise the use of number of processes available independent of the telescope's schedule. 

If we run the telescope every two weeks, we ensure that each process gets one batch of roughly the same size. And if we decide to change the schedule later on to e.g. weekly or monthly, each process still gets one batch of roughly the same size.

If the batch period is kept the same (for example one batch = one week), then this could be quite efficient when the schedule is set to bi-weekly, but if it is changed at one point to weekly or monthly, the batch size might be too small/big.

I think it comes down to how likely we'll think our schedules change, I thought a flexible batch size could account for this. If we're pretty set on keeping the schedule the same over time, it might be just overcomplicating things indeed!

I could add a variant of this (either with set batch period or flexible) to the telescope utils like Tuan did, I don't think the existing one is similar enough to what I'd need for crossref events (not using calender monthrange and I'd need a list of dates, instead of a list of periods)